metadata-store-directory-common/src/main/java/org/apache/helix/msdcommon/datamodel/TrieRoutingData.java

```diff
@@ -48,6 +48,10 @@ public TrieRoutingData(Map<String, List<String>> routingData)
       throw new InvalidRoutingDataException("routingData cannot be null or empty");
     }
 
+    if (!containsShardingKey(routingData)) {
+      throw new InvalidRoutingDataException("routingData needs at least 1 sharding key");
+    }
```

Comment:
Explain why this should be invalid? I thought we agreed that it's acceptable to have a realm without any keys.

Reply:
We agreed that it's okay for some realms to not have keys (which is handled correctly by skipping them), but not when **all** realms have no keys. The design of `TrieRoutingData` is "a set of data with routing capability", therefore empty routing data shouldn't be created; a group of realms with 0 key doesn't have routing capability either. 
In that situation, the current `TrieRoutingData` will create a structure in an erroneous state. The promise of `TrieRoutingData` is to contain at least 1 sharding key; with 0 sharding keys, some functionality will break. For example, `isShardingKeyInsertionValid("/")` called on a `TrieRoutingData` with 0 sharding key will return false, even though it should be true.

Reply:
So it's okay to have a bunch of realms with no keys as long as we have some realm with a key?

Reply:
Yes. The focus should be on "is there a sharding key". Without any sharding key the routing data is definitely invalid; with at least one it is valid.

Reply:
Okay. Is this clear in our documentation, both in code and user manual? We need to document this behavior so it's crystal clear to anyone that looks at this code.

Reply:
Also what would happen if a user initially creates N realms with keys, and ends up deleting all keys for the N realms? What happens when the user tries to delete the last remaining key?

Reply:
In my user manual I have briefly talked about invalid routing data. I will expand some sections for that. 
In terms of your second question, the refresh logic will trigger and `TrieRoutingData` will be removed from MSD when the last key is deleted, similar to how when the user deletes all routing data, `TrieRoutingData` is removed. Note that the raw routing data is not removed because the realms are still there. Also a reminder, when we don't have `TrieRoutingData`, the accessor behaves correctly.